---
title: "HistogramZoo"
author: "Helen Zhu & Stefan Eng"
output:
  rmarkdown:::html_vignette:
    toc: true
bibliography: >
  references.bib  
vignette: >
  %\VignetteIndexEntry{HistogramZoo}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning=FALSE, message=FALSE, width=500)
options(max.print=35)

library(HistogramZoo)

```

# Introduction
The HistogramZoo R package is a collection of functions which aim to identify and characterize patterns from histogram data.

# Quick Start

Begin with the generation of a `Histogram` object as a representation of the data.

The `segment_and_fit` function is a wrapper for a set of functions used for histogram characterization. Briefly, it identifies regions of increased density and summarizes it using statistical distributions. More details are provided below.

Finally, `summarize_results` returns a data.frame of segments and summary statistics and `create_coverageplot` returns a graphical representation of the results.

```{r basic_histogram_segmentation, include=FALSE}

library(HistogramZoo)

x = Histogram(
  histogram_data = c(0, 0, 1, 2, 3, 2, 1, 2, 3, 4, 5, 3, 1, 0),
  interval_start = 1:14,
  interval_end = 2:15,
  region_id = "my_histogram"
)

results = segment_and_fit(x, eps = 0.005)

results_table = summarize_results(results)

```

# Overview

## The `Histogram` and `GenomicHistogram` class

### Class-structure

A `Histogram` object contains 1 piece of essential data and 4 pieces of optional data:

* `histogram_data` is a numeric vector representing the histogram density

* `interval_start` and `histogram_end` are numeric vectors representing the start and end coordinates of each bin

* `bin_width` is a numeric indicator of the numeric length of bins. This can be automatically estimated from `interval_start` and `interval_end`

* `region_id` is a character id for the histogram

A `GenomicHistogram` object contains 2 additional pieces of optional data: `chr` for chromosome and `strand` for strand which is restricted to one of `+`, `-` and `*`. A `Histogram` and a `GenomicHistogram` differ in one of several key ways:

1. Conceptualization of bins: `Histogram` objects represent conventional histograms where data are binned on a real number scale. Starts are closed while ends are open with the exception of the last bin where both start and end are closed. `GenomicHistogram` objects were designed to represent a [Base 1](https://www.biostars.org/p/84686/) system which represent a set of indices on which observations are collected.

2. Restrictions on `bin_width`: `Histogram` objects require that the length of each bin is equal with the exception of the last bin which can be shorter than the rest. `GenomicHistogram` objects allow the lengths of each bin to vary as long as they are shorter or equal to the the designated `bin_width` and leaves the equal distribution of bin lengths up to user discretion. This allows users to construct `GenomicHistogram` objects using the `tile` and `binnedAverage` functions of `GenomicRanges`.

3. Disjoint intervals: Both `Histogram` and `GenomicHistogram` objects support the use of disjoint intervals but this feature was designed with `GenomicHistogram` objects in mind to support the representation of transcripts and genes as a set of exons which might not be consecutive intervals on the genomic coordinate system.

### Data loaders

Data loaders for `Histogram` objects:

* `observations_to_histogram` bins a set of observations at a designated `histogram_bin_width` to generate a `Histogram`

Data loaders for `GenomicHistogram` objects:

* `bigWig_to_histogram`

* `genome_BED_to_histogram`

* `transcript_BED_to_histogram`

## Segmentation and fitting distributions

### Finding local optima
* Discuss the three types of data properties we want to address

### The Fine-to-Coarse segmentation algorithm

The Fine-to-Coarse segmentation algorithm was described in @delon2005color.

## Filtering Low Entropy Regions

## Fitting Statistical Distributions

1. Truncated Distributions

2. Maximizing Uniform Elements

3. Metrics & Voting

## Exporting Results

1. Summarize results into a data.frame

2. Plotting

# References
