---
title: "HistogramZoo"
author: "Helen Zhu & Stefan Eng"
output:
  rmarkdown:::html_vignette:
    toc: true
bibliography: >
  references.bib  
vignette: >
  %\VignetteIndexEntry{HistogramZoo}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning=FALSE, message=FALSE, width=500)
options(max.print=35)

library(HistogramZoo)

```

# Introduction
The HistogramZoo R package is a collection of functions which aims to identify and characterize patterns from histogram data.

# Quick Start

A `Histogram` object contains 1 piece of essential data and 3 pieces of optional data:

* `histogram_data` is a numeric vector representing the histogram density

* `interval_start` and `histogram_end` are numeric vectors repesenting the start and end coordinates of each bin

* `region_id` is a character id for the histogram

The `segment_and_fit` function is a wrapper for a set of functions used for histogram characterization. Briefly, it identifies regions of increased density and summarizes it using statistical distributions. More details are provided below.

Finally, `summarize_results` returns a data.frame of segments and sumamry statistics and `create_coverageplot` returns a graphical representation of the results.

```{r basic_histogram_segmentation, include=FALSE}

library(HistogramZoo)

x = Histogram(
  histogram_data = c(0, 0, 1, 2, 3, 2, 1, 2, 3, 4, 5, 3, 1, 0),
  interval_start = 1:14,
  interval_end = 1:14,
  region_id = "my_histogram"
)

results = segment_and_fit(x, eps = 0.005)

results_table = summarize_results(results)

```

# Overview

## The `Histogram` and `GenomicHistogram` class

1. Class-structure

2. Data loaders

3. A note about data types.

## Fine-to-Coarse Segmentation Algorithm

1. Finding local optima

2. Tuning epsilon

## Filtering Low Entropy Regions

## Fitting Statistical Distributions

1. Truncated Distributions

2. Maximizing Uniform Elements

3. Metrics & Voting

## Exporting Results

1. Summarize results into a data.frame

2. Plotting

# Applications in Genomics

## Example 1: Identifying consensus regions across transcriptome BED files
A common step in the bioinformatic analysis of epigenetic pulldown experiments such as ChIP-seq, ATAC-seq, MeRIP-seq, CLIP-seq and other methods is to call peaks across biological replicates. In such studies, it is often desirable to identify a set of common peaks between replicates, count reads and conduct differential analysis. It is easy to identify common peaks with few replicates by simply taking the union of the set of peak regions. However, as datasets grow, the length of peak unions also grow, spanning the finite space defined by the genome or transcriptome. Additionally, epitranscriptomic pulldown experiments are growing in popularity. While peak callers for the transcriptome (e.g. exomePeak, MeTPeak) have been created, there is no available tool to identify overlapping peaks that span multiple exons. HistogramZoo is a tool that implements several models described in the literature to identify a consensus set of regions for RNA pulldown experiments.

```{r consensus_peaks_transcriptome, include=FALSE}

# Load Data
x = Histogram(c(0, 0, 1, 2, 3, 2, 1, 2, 3, 4, 5, 3, 1, 0))

results = segment_and_fit(x, eps = 0.005)

results_table = summarize_results(results)

```

## Example 2: Fine tuning peaks using coverage signals on bigWig files

# References
