<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Function reference â€¢ HistogramZoo</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">HistogramZoo</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.3.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/HistogramZoo.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Function reference</h1>
    </div>

    <div class="section level2">
      <h2 id="all-functions">All functions<a class="anchor" aria-label="anchor" href="#all-functions"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="GTF_to_GRangesList.html">GTF_to_GRangesList()</a></code> 
        </dt>
        <dd>Produces a GRangesList out of a GTF file, each element represents the exons of a gene or transcript</dd>
      </dl><dl><dt>
          
          <code><a href="GenomicHistogram.html">GenomicHistogram()</a></code> 
        </dt>
        <dd>Generates an S3 `GenomicHistogram` object</dd>
      </dl><dl><dt>
          
          <code><a href="Histogram.html">Histogram()</a></code> 
        </dt>
        <dd>Generates an S3 `Histogram` object</dd>
      </dl><dl><dt>
          
          <code><a href="are_colours.html">are_colours()</a></code> 
        </dt>
        <dd>Verify if a vector contains R colours</dd>
      </dl><dl><dt>
          
          <code><a href="base0_to_base1.html">base0_to_base1()</a></code> 
        </dt>
        <dd>Converts a GRanges object from base 0 to base 1</dd>
      </dl><dl><dt>
          
          <code><a href="bigWig_to_histogram.html">bigWig_to_histogram()</a></code> 
        </dt>
        <dd>Generates GenomicHistogram objects from the coverage of a bigWig file</dd>
      </dl><dl><dt>
          
          <code><a href="bin_log_likelihood.html">bin_log_likelihood()</a></code> 
        </dt>
        <dd>Computes the negative log likelihood of an underlying continuous distribution from binned data.</dd>
      </dl><dl><dt>
          
          <code><a href="calculate_probability_difference.html">calculate_probability_difference()</a></code> 
        </dt>
        <dd>Calculate whether h[a,b] &gt; p[a,b]</dd>
      </dl><dl><dt>
          
          <code><a href="compute_coverage_on_bins.html">compute_coverage_on_bins()</a></code> 
        </dt>
        <dd>A helper function for coverage_to_histogram that returns a dataframe from a set of ranges</dd>
      </dl><dl><dt>
          
          <code><a href="correct_fitted_value.html">correct_fitted_value()</a></code> 
        </dt>
        <dd>Correcting for Jaccard/Intersection</dd>
      </dl><dl><dt>
          
          <code><a href="coverage_to_histogram.html">coverage_to_histogram()</a></code> 
        </dt>
        <dd>Generates a GenomicHistogram object from coverage data</dd>
      </dl><dl><dt>
          
          <code><a href="create_coverageplot.html">create_coverageplot()</a></code> 
        </dt>
        <dd>create_coverageplot creates a histogram/coverage plot with the potential to add annotations</dd>
      </dl><dl><dt>
          
          <code><a href="create_histogram.html">create_histogram()</a></code> 
        </dt>
        <dd>Creates a histogram with potentially variable length bins</dd>
      </dl><dl><dt>
          
          <code><a href="create_layerplot.html">create_layerplot()</a></code> 
        </dt>
        <dd>Stacks together multiple plots</dd>
      </dl><dl><dt>
          
          <code><a href="create_residualplot.html">create_residualplot()</a></code> 
        </dt>
        <dd>create_residualplot creates a residual plot between fitted and observed data</dd>
      </dl><dl><dt>
          
          <code><a href="create_trackplot.html">create_trackplot()</a></code> 
        </dt>
        <dd>create_trackplot creates a horizontal plot with rows representing separate tracks</dd>
      </dl><dl><dt>
          
          <code><a href="dgamma_flip.html">dgamma_flip()</a></code> 
        </dt>
        <dd>gamma flip</dd>
      </dl><dl><dt>
          
          <code><a href="dput_str.html">dput_str()</a></code> 
        </dt>
        <dd>Return a string representation of an object</dd>
      </dl><dl><dt>
          
          <code><a href="dtgamma.html">dtgamma()</a></code> 
        </dt>
        <dd>truncated gamma</dd>
      </dl><dl><dt>
          
          <code><a href="dtgamma_flip.html">dtgamma_flip()</a></code> 
        </dt>
        <dd>truncated gamma flip</dd>
      </dl><dl><dt>
          
          <code><a href="extract_segments.html">extract_segments()</a></code> 
        </dt>
        <dd>Represent a single segment as a set of intervals</dd>
      </dl><dl><dt>
          
          <code><a href="extract_stats_from_models.html">extract_stats_from_models()</a></code> 
        </dt>
        <dd>Extract stats from models</dd>
      </dl><dl><dt>
          
          <code><a href="find_all_meaningful_gap.html">find_all_meaningful_gap()</a></code> 
        </dt>
        <dd>Find all meaningful gaps</dd>
      </dl><dl><dt>
          
          <code><a href="find_change_points.html">find_change_points()</a></code> 
        </dt>
        <dd>Find changepoints in a vector with uniform stretches of values</dd>
      </dl><dl><dt>
          
          <code><a href="find_consecutive_threshold.html">find_consecutive_threshold()</a></code> 
        </dt>
        <dd>Returns the indices for consecutive elements of a vector that are greater than a specified threshold</dd>
      </dl><dl><dt>
          
          <code><a href="find_consensus_model.html">find_consensus_model()</a></code> 
        </dt>
        <dd>Methods for voting for a consensus model based on the metrics of fit_distributions</dd>
      </dl><dl><dt>
          
          <code><a href="find_local_optima.html">find_local_optima()</a></code> 
        </dt>
        <dd>find_local_optima returns the local optima of histograms</dd>
      </dl><dl><dt>
          
          <code><a href="fit_distributions.html">fit_distributions()</a></code> 
        </dt>
        <dd>Fit the model parameters by optimizing a histogram metric</dd>
      </dl><dl><dt>
          
          <code><a href="fit_uniform.numeric.html">fit_uniform(<i>&lt;numeric&gt;</i>)</a></code> 
        </dt>
        <dd>Fit a uniform distribution to a histogram</dd>
      </dl><dl><dt>
          
          <code><a href="format_print_intervals.html">format_print_intervals()</a></code> 
        </dt>
        <dd>Formats print intervals with introns and one-bp bins</dd>
      </dl><dl><dt>
          
          <code><a href="ftc.html">ftc()</a></code> 
        </dt>
        <dd>Fine-to-coarse segmentation algorithm</dd>
      </dl><dl><dt>
          
          <code><a href="generate_granges_from_parent_and_ranges.html">generate_granges_from_parent_and_ranges()</a></code> 
        </dt>
        <dd>A helper function for coverage_to_histogram that returns a dataframe from a set of ranges</dd>
      </dl><dl><dt>
          
          <code><a href="generate_identifiers.html">generate_identifiers()</a></code> 
        </dt>
        <dd>Generates region_ids for GRanges objects</dd>
      </dl><dl><dt>
          
          <code><a href="generate_row_ids.html">generate_row_ids()</a></code> 
        </dt>
        <dd>Generates row names from row data</dd>
      </dl><dl><dt>
          
          <code><a href="generate_uniform_distribution.html">generate_uniform_distribution()</a></code> 
        </dt>
        <dd>Uniform density generation</dd>
      </dl><dl><dt>
          
          <code><a href="generate_xlabels.html">generate_xlabels()</a></code> 
        </dt>
        <dd>Generates x labels for histogram-based plots, `create_coverageplot` and `create_residualplot`</dd>
      </dl><dl><dt>
          
          <code><a href="genome_BED_to_histogram.html">genome_BED_to_histogram()</a></code> 
        </dt>
        <dd>Generate GenomicHistogram objects from BED files</dd>
      </dl><dl><dt>
          
          <code><a href="grenader.html">grenader()</a></code> 
        </dt>
        <dd>Calculate the Grenander estimator of a given density vector</dd>
      </dl><dl><dt>
          
          <code><a href="histogram.chisq.html">histogram.chisq()</a></code> 
        </dt>
        <dd>Chi-squared</dd>
      </dl><dl><dt>
          
          <code><a href="histogram.intersection.html">histogram.intersection()</a></code> 
        </dt>
        <dd>Histogram intersection</dd>
      </dl><dl><dt>
          
          <code><a href="histogram.jaccard.html">histogram.jaccard()</a></code> 
        </dt>
        <dd>Jaccard index</dd>
      </dl><dl><dt>
          
          <code><a href="histogram.ks.html">histogram.ks()</a></code> 
        </dt>
        <dd>Kolmogorov-Smirnov divergence</dd>
      </dl><dl><dt>
          
          <code><a href="histogram.mse.html">histogram.mse()</a></code> 
        </dt>
        <dd>Mean-squared error</dd>
      </dl><dl><dt>
          
          <code><a href="identify_uniform_segment.html">identify_uniform_segment()</a></code> 
        </dt>
        <dd>Finds the largest uniform segment that is longer than threshold</dd>
      </dl><dl><dt>
          
          <code><a href="index_to_start_end.html">index_to_start_end()</a></code> 
        </dt>
        <dd>Convert a vector of points into a data.frame of start/end points representing
disjoint intervals</dd>
      </dl><dl><dt>
          
          <code><a href="is_equal_integer.html">is_equal_integer()</a></code> 
        </dt>
        <dd>Checks if a numeric vector can be used as an integer vector</dd>
      </dl><dl><dt>
          
          <code><a href="labels_helper.html">labels_helper()</a></code> 
        </dt>
        <dd>A helper function that generates labels at specific Histogram bins</dd>
      </dl><dl><dt>
          
          <code><a href="maximal_meaningful_interval.html">maximal_meaningful_interval()</a></code> 
        </dt>
        <dd>Identifies the maximum meaningful interval from a set of overlapping
intervals, can also be applied to find the max gap</dd>
      </dl><dl><dt>
          
          <code><a href="maximum_entropy.html">maximum_entropy()</a></code> 
        </dt>
        <dd>Computes H, the maximum H_h,p([a,b])</dd>
      </dl><dl><dt>
          
          <code><a href="meaningful_gap.html">meaningful_gap()</a></code> 
        </dt>
        <dd>Determines whether h[a,b] is a meaningful gap</dd>
      </dl><dl><dt>
          
          <code><a href="meaningful_gaps_local.html">meaningful_gaps_local()</a></code> 
        </dt>
        <dd>Finds the meaningful gaps between the points in s</dd>
      </dl><dl><dt>
          
          <code><a href="meaningful_interval.html">meaningful_interval()</a></code> 
        </dt>
        <dd>Determines whether h[a,b] is a meaningful interval</dd>
      </dl><dl><dt>
          
          <code><a href="monotone_cost.html">monotone_cost()</a></code> 
        </dt>
        <dd>Compute the monotone cost,</dd>
      </dl><dl><dt>
          
          <code><a href="new_GenomicHistogram.html">new_GenomicHistogram()</a></code> 
        </dt>
        <dd>GenomicHistogram constructor</dd>
      </dl><dl><dt>
          
          <code><a href="new_Histogram.html">new_Histogram()</a></code> 
        </dt>
        <dd>Constructs a new Histogram object</dd>
      </dl><dl><dt>
          
          <code><a href="observations_to_histogram.html">observations_to_histogram()</a></code> 
        </dt>
        <dd>Take a vector of values and get the histogram for integer breaks</dd>
      </dl><dl><dt>
          
          <code><a href="pgamma_flip.html">pgamma_flip()</a></code> 
        </dt>
        <dd>gamma flip</dd>
      </dl><dl><dt>
          
          <code><a href="ptgamma.html">ptgamma()</a></code> 
        </dt>
        <dd>truncated gamma</dd>
      </dl><dl><dt>
          
          <code><a href="ptgamma_flip.html">ptgamma_flip()</a></code> 
        </dt>
        <dd>truncated gamma flip</dd>
      </dl><dl><dt>
          
          <code><a href="qgamma_flip.html">qgamma_flip()</a></code> 
        </dt>
        <dd>gamma flip</dd>
      </dl><dl><dt>
          
          <code><a href="qtgamma.html">qtgamma()</a></code> 
        </dt>
        <dd>truncated gamma</dd>
      </dl><dl><dt>
          
          <code><a href="qtgamma_flip.html">qtgamma_flip()</a></code> 
        </dt>
        <dd>truncated gamma</dd>
      </dl><dl><dt>
          
          <code><a href="reassign_region_id.html">reassign_region_id()</a></code> 
        </dt>
        <dd>reassign_region_id for Histogram objects</dd>
      </dl><dl><dt>
          
          <code><a href="relative_entropy.html">relative_entropy()</a></code> 
        </dt>
        <dd>Kullback-Leibler divergence (Relative Entropy)</dd>
      </dl><dl><dt>
          
          <code><a href="remove_max_gaps.html">remove_max_gaps()</a></code> 
        </dt>
        <dd>Creates a new set of segments from a partition of points</dd>
      </dl><dl><dt>
          
          <code><a href="return_x_points.html">return_x_points()</a></code> 
        </dt>
        <dd>Returns the segment_and_fit x coordinates of the identified points</dd>
      </dl><dl><dt>
          
          <code><a href="return_y_points.html">return_y_points()</a></code> 
        </dt>
        <dd>Returns the segment_and_fit y coordinates of the identified points</dd>
      </dl><dl><dt>
          
          <code><a href="rgamma_flip.html">rgamma_flip()</a></code> 
        </dt>
        <dd>gamma flip</dd>
      </dl><dl><dt>
          
          <code><a href="rtgamma.html">rtgamma()</a></code> 
        </dt>
        <dd>truncated gamma</dd>
      </dl><dl><dt>
          
          <code><a href="rtgamma_flip.html">rtgamma_flip()</a></code> 
        </dt>
        <dd>truncated gamma flip</dd>
      </dl><dl><dt>
          
          <code><a href="scale_model_params.html">scale_model_params()</a></code> 
        </dt>
        <dd>Rescale parameter statistics based on original fit bin width</dd>
      </dl><dl><dt>
          
          <code><a href="segment_and_fit.html">segment_and_fit()</a></code> 
        </dt>
        <dd>Segmentation of histograms and distribution fitting</dd>
      </dl><dl><dt>
          
          <code><a href="summarize_results.html">summarize_results()</a></code> 
        </dt>
        <dd>Formats results of segment_and_fit</dd>
      </dl><dl><dt>
          
          <code><a href="transcript_BED_to_histogram.html">transcript_BED_to_histogram()</a></code> 
        </dt>
        <dd>Generate GenomicHistogram objects from BED files using coordinates defined by a GTF file</dd>
      </dl><dl><dt>
          
          <code><a href="validate_GenomicHistogram.html">validate_GenomicHistogram()</a></code> 
        </dt>
        <dd>Validates GenomicHistogram objects</dd>
      </dl><dl><dt>
          
          <code><a href="validate_Histogram.html">validate_Histogram()</a></code> 
        </dt>
        <dd>Validates Histogram objects</dd>
      </dl><dl><dt>
          
          <code><a href="weighted.moments.html">weighted.sd()</a></code> <code><a href="weighted.moments.html">weighted.var()</a></code> <code><a href="weighted.moments.html">weighted.skewness()</a></code> <code><a href="weighted.moments.html">weighted.mean(<i>&lt;Histogram&gt;</i>)</a></code> 
        </dt>
        <dd>Weighted higher order moments</dd>
      </dl><dl><dt>
          
          <code><a href="weighted_coverage.html">weighted_coverage()</a></code> 
        </dt>
        <dd>A helper function for coverage_to_histogram that returns the weighted coverage of a disjoint bin</dd>
      </dl></div>
  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Helen Zhu, Stefan Eng.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

