<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="HistogramZoo">
<title>HistogramZoo • HistogramZoo</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="HistogramZoo">
<meta property="og:description" content="HistogramZoo">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">HistogramZoo</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.3.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/HistogramZoo.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>HistogramZoo</h1>
                        <h4 data-toc-skip class="author">Helen Zhu &amp;
Stefan Eng</h4>
            
      
      
      <div class="d-none name"><code>HistogramZoo.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The HistogramZoo R package is a collection of functions which aim to
identify and characterize patterns from histogram data.</p>
</div>
<div class="section level2">
<h2 id="the-histogram-and-genomichistogram-class">The Histogram and GenomicHistogram class<a class="anchor" aria-label="anchor" href="#the-histogram-and-genomichistogram-class"></a>
</h2>
<p>A <strong>Histogram</strong> object contains 1 piece of essential
data and 4 pieces of optional data:</p>
<ul>
<li><p><strong>histogram_data</strong> is a numeric vector representing
the histogram density or counts</p></li>
<li><p><strong>interval_start</strong> and
<strong>histogram_end</strong> are numeric vectors representing the
start and end coordinates of each bin</p></li>
<li><p><strong>bin_width</strong> is the numeric length of bins. This
can be automatically estimated from <code>interval_start</code> and
<code>interval_end</code></p></li>
<li><p><strong>region_id</strong> is a character id for the
histogram</p></li>
</ul>
<p>A <strong>GenomicHistogram</strong> object contains 2 additional
pieces of optional data: <strong>chr</strong> for chromosome and
<strong>strand</strong> for strand which is restricted to one of
<code>+</code>, <code>-</code> and <code>*</code>. A Histogram and a
GenomicHistogram differ in one of several ways:</p>
<ol style="list-style-type: decimal">
<li><p>Conceptualization of bins: <code>Histogram</code> objects
represent conventional histograms where data are binned on a real number
scale. Starts are closed while ends are open with the exception of the
last bin where both start and end are closed.
<code>GenomicHistogram</code> objects were designed to represent a
discrete <a href="https://www.biostars.org/p/84686/" class="external-link">base 1</a> system
where each histogram bin represents a set of indices on which
observations are collected.</p></li>
<li><p>Restrictions on <code>bin_width</code>: <code>Histogram</code>
objects require that the length of each bin is equal with the exception
of the last bin which can be shorter than the rest.
<code>GenomicHistogram</code> objects allow the lengths of each bin to
vary as long as they are shorter or equal to the the designated
<code>bin_width</code> and leaves the equal distribution of bin lengths
up to user discretion. This allows users to construct
<code>GenomicHistogram</code> objects using the <code>tile</code> and
<code>binnedAverage</code> functions of
<code>GenomicRanges</code>.</p></li>
<li><p>Disjoint intervals: Both <code>Histogram</code> and
<code>GenomicHistogram</code> objects support the use of disjoint
intervals but this feature was designed with
<code>GenomicHistogram</code> objects in mind to support the
representation of transcripts and genes as a set of exons which might
not be consecutive intervals on the genomic coordinate system.</p></li>
</ol>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_histogram</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Histogram.html">Histogram</a></span><span class="op">(</span></span>
<span>  histogram_data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">3</span>,<span class="fl">1</span>,<span class="fl">5</span>,<span class="fl">6</span><span class="op">)</span>,</span>
<span>  interval_start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">3</span>,<span class="fl">4</span>,<span class="fl">5</span><span class="op">)</span>,</span>
<span>  interval_end <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">3</span>,<span class="fl">4</span>,<span class="fl">5</span>,<span class="fl">6</span><span class="op">)</span>,</span>
<span>  bin_width <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  region_id <span class="op">=</span> <span class="st">"my_histogram"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/create_coverageplot.html">create_coverageplot</a></span><span class="op">(</span></span>
<span>  <span class="va">my_histogram</span>,</span>
<span>  main.cex <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  xaxis.tck <span class="op">=</span> <span class="fl">0</span>, </span>
<span>  yaxis.tck <span class="op">=</span> <span class="fl">0</span>, </span>
<span>  xlab.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  ylab.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  xaxis.cex <span class="op">=</span> <span class="fl">0.8</span>, </span>
<span>  yaxis.cex <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  xat <span class="op">=</span> <span class="fu"><a href="../reference/generate_xlabels.html">generate_xlabels</a></span><span class="op">(</span><span class="va">my_histogram</span>, n_labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">my_histogram</span><span class="op">)</span>, return_xat <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>, </span>
<span>  xaxis.lab <span class="op">=</span> <span class="fu"><a href="../reference/generate_xlabels.html">generate_xlabels</a></span><span class="op">(</span><span class="va">my_histogram</span>, n_labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">my_histogram</span><span class="op">)</span><span class="op">)</span>, </span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">my_genomic_histogram</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GenomicHistogram.html">GenomicHistogram</a></span><span class="op">(</span></span>
<span>  histogram_data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4</span>,<span class="fl">5</span>,<span class="fl">2</span>,<span class="fl">3</span>,<span class="fl">6</span>,<span class="fl">8</span><span class="op">)</span>,</span>
<span>  interval_start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">6</span><span class="op">)</span>,</span>
<span>  interval_end <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">6</span><span class="op">)</span>,</span>
<span>  bin_width <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  region_id <span class="op">=</span> <span class="st">"my_genomic_histogram"</span>,</span>
<span>  chr <span class="op">=</span> <span class="st">"chr1"</span>,</span>
<span>  strand <span class="op">=</span> <span class="st">"*"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/create_coverageplot.html">create_coverageplot</a></span><span class="op">(</span></span>
<span>  <span class="va">my_genomic_histogram</span>,</span>
<span>  main.cex <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  xaxis.tck <span class="op">=</span> <span class="fl">0</span>, </span>
<span>  yaxis.tck <span class="op">=</span> <span class="fl">0</span>, </span>
<span>  xlab.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  ylab.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  xaxis.cex <span class="op">=</span> <span class="fl">0.8</span>, </span>
<span>  yaxis.cex <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  xat <span class="op">=</span> <span class="fu"><a href="../reference/generate_xlabels.html">generate_xlabels</a></span><span class="op">(</span><span class="va">my_genomic_histogram</span>, n_labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">my_genomic_histogram</span><span class="op">)</span>, return_xat <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>, </span>
<span>  xaxis.lab <span class="op">=</span> <span class="fu"><a href="../reference/generate_xlabels.html">generate_xlabels</a></span><span class="op">(</span><span class="va">my_genomic_histogram</span>, n_labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">my_genomic_histogram</span><span class="op">)</span><span class="op">)</span>, </span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="HistogramZoo_files/figure-html/Histogram_example-1.png" width="40%"><img src="HistogramZoo_files/figure-html/Histogram_example-2.png" width="40%"></p>
</div>
<div class="section level2">
<h2 id="data-loaders">Data loaders<a class="anchor" aria-label="anchor" href="#data-loaders"></a>
</h2>
<div class="section level3">
<h3 id="data-loaders-for-histogram-objects">Data loaders for <code>Histogram</code> objects<a class="anchor" aria-label="anchor" href="#data-loaders-for-histogram-objects"></a>
</h3>
<p><strong>observations_to_histogram</strong> bins a set of observations
at a designated <code>histogram_bin_width</code> to generate a
<code>Histogram</code></p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">314</span><span class="op">)</span></span>
<span><span class="va">my_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">10000</span>, mean <span class="op">=</span> <span class="fl">10</span>, sd <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">my_histogram</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/observations_to_histogram.html">observations_to_histogram</a></span><span class="op">(</span><span class="va">my_data</span>, histogram_bin_width <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/create_coverageplot.html">create_coverageplot</a></span><span class="op">(</span></span>
<span>  <span class="va">my_histogram</span>, </span>
<span>  main.cex <span class="op">=</span> <span class="cn">NULL</span>, </span>
<span>  xaxis.tck <span class="op">=</span> <span class="fl">0</span>, </span>
<span>  yaxis.tck <span class="op">=</span> <span class="fl">0</span>, </span>
<span>  xlab.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  ylab.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  xaxis.cex <span class="op">=</span> <span class="fl">0.8</span>, </span>
<span>  yaxis.cex <span class="op">=</span> <span class="fl">0.8</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="HistogramZoo_files/figure-html/observations_to_histogram-1.png" width="50%"></p>
</div>
<div class="section level3">
<h3 id="data-loaders-for-genomichistogram-objects">Data loaders for <code>GenomicHistogram</code> objects<a class="anchor" aria-label="anchor" href="#data-loaders-for-genomichistogram-objects"></a>
</h3>
<p><strong>bigWig_to_histogram</strong> bins the coverage from a single
bigWig file on a set of user-defined regions (provided as a GRanges
object, a GRangesList* object or a GTF file) and returns a list of
GenomicHistogram objects.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges" class="external-link">GenomicRanges</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">filename</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"bigwigs"</span>,  <span class="st">"S1.bw"</span>, package <span class="op">=</span> <span class="st">"HistogramZoo"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">regions</span> <span class="op">&lt;-</span> <span class="fu">GenomicRanges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span></span>
<span>  seqnames <span class="op">=</span> <span class="st">"chr1"</span>,</span>
<span>  <span class="fu">IRanges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span></span>
<span>    start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">17950</span>, <span class="fl">19350</span><span class="op">)</span>,</span>
<span>    end <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">18000</span>, <span class="fl">19600</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  strand <span class="op">=</span> <span class="st">"*"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">bigwig_histograms</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bigWig_to_histogram.html">bigWig_to_histogram</a></span><span class="op">(</span></span>
<span>  filename <span class="op">=</span> <span class="va">filename</span>,</span>
<span>  regions <span class="op">=</span> <span class="va">regions</span>,</span>
<span>  histogram_bin_size <span class="op">=</span> <span class="fl">10</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><strong>genome_BED_to_histogram</strong> bins the coverage from a set
of genomic BED files and returns a list of GenomicHistogram objects.
Users have the option of providing a set of regions as a GRanges or
GRangesList* object.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">datadir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"dna_bedfiles"</span>,  package <span class="op">=</span> <span class="st">"HistogramZoo"</span><span class="op">)</span></span>
<span><span class="va">genome_bed_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">datadir</span>, pattern <span class="op">=</span> <span class="st">".bed$"</span><span class="op">)</span></span>
<span><span class="va">genome_bed_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">datadir</span>, <span class="va">genome_bed_files</span><span class="op">)</span></span>
<span></span>
<span><span class="va">genome_bed_histograms</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/genome_BED_to_histogram.html">genome_BED_to_histogram</a></span><span class="op">(</span></span>
<span>  filenames <span class="op">=</span> <span class="va">genome_bed_files</span>,</span>
<span>  n_fields <span class="op">=</span> <span class="fl">6</span>,</span>
<span>  histogram_bin_size <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><strong>transcript_BED_to_histogram</strong>, to account for
overlapping transcript annotations, bins the coverage on a set of BED
files where each element can be pre-assigned to a region using the 4th
column (name) of the BED file.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">datadir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"rna_bedfiles"</span>, package <span class="op">=</span> <span class="st">"HistogramZoo"</span><span class="op">)</span></span>
<span><span class="va">transcript_bed_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">datadir</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Sample."</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">20</span>, <span class="st">".bed"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">gtf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"genes.gtf"</span>, package <span class="op">=</span> <span class="st">"HistogramZoo"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">histograms</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/transcript_BED_to_histogram.html">transcript_BED_to_histogram</a></span><span class="op">(</span></span>
<span>  filenames <span class="op">=</span> <span class="va">transcript_bed_files</span>,</span>
<span>  n_fields <span class="op">=</span> <span class="fl">12</span>,</span>
<span>  gtf <span class="op">=</span> <span class="va">gtf</span>,</span>
<span>  gene_or_transcript <span class="op">=</span> <span class="st">"gene"</span>,</span>
<span>  histogram_bin_size <span class="op">=</span> <span class="fl">10</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="notes-on-common-parameters">Notes on common parameters<a class="anchor" aria-label="anchor" href="#notes-on-common-parameters"></a>
</h4>
<ul>
<li><p><strong>regions</strong> <code>[bigWig, genomic BED]</code>:
Using a GRangesList object allows the assignment of disjoint intervals
to a GenomicHistogram (e.g. representing a set of exons that can be
joined to form a mature transcript) where each GRanges object in the
GRangesList represents the elements of a single Histogram. Using a
GRanges object creates a separate GenomicHistogram for each element in
the GRanges object.</p></li>
<li><p><strong>histogram_bin_size</strong>
<code>[bigWig, genomic and transcriptomic BED]</code>: Bins are
generated using <code><a href="https://rdrr.io/pkg/IRanges/man/IPosRanges-class.html" class="external-link">GenomicRanges::tile</a></code> and coverage is
calculated using <code><a href="https://rdrr.io/pkg/GenomicRanges/man/genomicvars.html" class="external-link">GenomicRanges::binnedAverage</a></code>. This
strategy strives to create bins of equal size less than or equal to the
designated <code>histogram_bin_size</code> (see above for discussion of
the difference in <code>bin_width</code> restriction.)</p></li>
<li><p><strong>allow_overlapping_segments_per_sample</strong>
<code>[transcriptomic and genomic BED]</code>: Each BED file is assumed
to contain the elements of a single experimental sample. Thus,
overlapping elements in each BED file can be eliminated by setting the
parameter to <code>FALSE</code>.</p></li>
</ul>
</div>
</div>
</div>
<div class="section level2">
<h2 id="segmentation-and-distribution-fitting">Segmentation and Distribution Fitting<a class="anchor" aria-label="anchor" href="#segmentation-and-distribution-fitting"></a>
</h2>
<p>The function <strong>segment_and_fit</strong> is designed to be a
pipeline constructed from the following sub-functions which enables the
segmentation, denoising and fitting of statistical distributions on
Histogram and GenomicHistogram objects. Each of the parameters in
<code>segment_and_fit</code> have counterparts in the sub-functions
which is explored in further detail below.</p>
<div class="section level3">
<h3 id="finding-local-optima">Finding local optima<a class="anchor" aria-label="anchor" href="#finding-local-optima"></a>
</h3>
<p><strong>find_local_optima</strong> identifies the local minima and
maxima on a vector of counts (histogram data).</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Generating a Histogram </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">314</span><span class="op">)</span></span>
<span><span class="va">dt</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">10000</span>, mean <span class="op">=</span> <span class="fl">20</span>, sd <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">xhist</span> <span class="op">=</span> <span class="fu"><a href="../reference/observations_to_histogram.html">observations_to_histogram</a></span><span class="op">(</span><span class="va">dt</span>, histogram_bin_width <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">data</span> <span class="op">=</span> <span class="va">xhist</span><span class="op">$</span><span class="va">histogram_data</span></span>
<span></span>
<span><span class="co"># basic use</span></span>
<span><span class="va">optima</span> <span class="op">=</span> <span class="fu"><a href="../reference/find_local_optima.html">find_local_optima</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span>
<span><span class="va">optima</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">optima</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/create_coverageplot.html">create_coverageplot</a></span><span class="op">(</span></span>
<span>  <span class="va">xhist</span>, </span>
<span>  add.points <span class="op">=</span> <span class="cn">T</span>, </span>
<span>  points.x <span class="op">=</span> <span class="va">optima</span>, </span>
<span>  points.y <span class="op">=</span> <span class="va">data</span><span class="op">[</span><span class="va">optima</span><span class="op">]</span>, </span>
<span>  points.col <span class="op">=</span> <span class="st">"red"</span>,</span>
<span>  main <span class="op">=</span> <span class="st">"basic optima"</span>,</span>
<span>  main.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  xaxis.tck <span class="op">=</span> <span class="fl">0</span>, </span>
<span>  yaxis.tck <span class="op">=</span> <span class="fl">0</span>, </span>
<span>  xlab.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  ylab.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  xaxis.cex <span class="op">=</span> <span class="fl">0.8</span>, </span>
<span>  yaxis.cex <span class="op">=</span> <span class="fl">0.8</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="HistogramZoo_files/figure-html/find_local_optima_basic-1.png" width="50%"></p>
<p>The parameter <code>threshold</code> allows the user to tune the
sparsity of local optima to histograms with potentially noisy counts,
i.e. by thresholding the difference between neighbouring optima.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Generating a Histogram with noisy counts</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">314</span><span class="op">)</span></span>
<span><span class="va">dt</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">1000</span>, mean <span class="op">=</span> <span class="fl">20</span>, sd <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">xhist</span> <span class="op">=</span> <span class="fu"><a href="../reference/observations_to_histogram.html">observations_to_histogram</a></span><span class="op">(</span><span class="va">dt</span>, histogram_bin_width <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">data</span> <span class="op">=</span> <span class="va">xhist</span><span class="op">$</span><span class="va">histogram_data</span></span>
<span></span>
<span><span class="co"># find_local_optima with threshold 0</span></span>
<span><span class="va">zero_threshold_optima</span> <span class="op">=</span> <span class="fu"><a href="../reference/find_local_optima.html">find_local_optima</a></span><span class="op">(</span><span class="va">data</span>, threshold <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">zero_threshold_optima</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">zero_threshold_optima</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/create_coverageplot.html">create_coverageplot</a></span><span class="op">(</span></span>
<span>  <span class="va">xhist</span>,</span>
<span>  add.points <span class="op">=</span> <span class="cn">T</span>, </span>
<span>  points.x <span class="op">=</span> <span class="va">zero_threshold_optima</span>,</span>
<span>  points.y <span class="op">=</span> <span class="va">data</span><span class="op">[</span><span class="va">zero_threshold_optima</span><span class="op">]</span>,</span>
<span>  points.col <span class="op">=</span> <span class="st">"red"</span>, </span>
<span>  type <span class="op">=</span> <span class="st">"h"</span>,</span>
<span>  main <span class="op">=</span> <span class="st">"threshold = 0"</span>,</span>
<span>  main.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  xaxis.tck <span class="op">=</span> <span class="fl">0</span>, </span>
<span>  yaxis.tck <span class="op">=</span> <span class="fl">0</span>, </span>
<span>  xlab.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  ylab.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  xaxis.cex <span class="op">=</span> <span class="fl">0.8</span>, </span>
<span>  yaxis.cex <span class="op">=</span> <span class="fl">0.8</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># find_local_optima with threshold 2</span></span>
<span><span class="va">threshold_optima</span> <span class="op">=</span> <span class="fu"><a href="../reference/find_local_optima.html">find_local_optima</a></span><span class="op">(</span><span class="va">data</span>, threshold <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">threshold_optima</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">threshold_optima</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/create_coverageplot.html">create_coverageplot</a></span><span class="op">(</span></span>
<span>  <span class="va">xhist</span>, </span>
<span>  add.points <span class="op">=</span> <span class="cn">T</span>, </span>
<span>  points.x <span class="op">=</span> <span class="va">threshold_optima</span>, </span>
<span>  points.y <span class="op">=</span> <span class="va">data</span><span class="op">[</span><span class="va">threshold_optima</span><span class="op">]</span>, </span>
<span>  points.col <span class="op">=</span> <span class="st">"red"</span>, </span>
<span>  type <span class="op">=</span> <span class="st">"h"</span>,</span>
<span>  main <span class="op">=</span> <span class="st">"threshold = 2"</span>,</span>
<span>  main.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  xaxis.tck <span class="op">=</span> <span class="fl">0</span>, </span>
<span>  yaxis.tck <span class="op">=</span> <span class="fl">0</span>, </span>
<span>  xlab.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  ylab.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  xaxis.cex <span class="op">=</span> <span class="fl">0.8</span>, </span>
<span>  yaxis.cex <span class="op">=</span> <span class="fl">0.8</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="HistogramZoo_files/figure-html/find_local_optima_threshold-1.png" width="40%"><img src="HistogramZoo_files/figure-html/find_local_optima_threshold-2.png" width="40%"></p>
<p>The parameter <code>flat_endpoints</code> allows the user to better
fit histograms with stretches of equal values (e.g. pile-ups of BED
files in a GenomicHistogram). Setting the parameter to <code>TRUE</code>
will allow the user to return both endpoints of a flat region while
setting the parameter to <code>FALSE</code> will return the estimated
midpoint of a flat region.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Generating a Histogram with consecutive equal values</span></span>
<span><span class="va">data</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">4</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">3</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">1</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">xhist</span> <span class="op">=</span> <span class="fu"><a href="../reference/Histogram.html">Histogram</a></span><span class="op">(</span>histogram_data <span class="op">=</span> <span class="va">data</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># flat_endpoints</span></span>
<span><span class="va">optima_flat</span> <span class="op">=</span> <span class="fu"><a href="../reference/find_local_optima.html">find_local_optima</a></span><span class="op">(</span><span class="va">data</span>, flat_endpoints <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">optima_flat</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">optima_flat</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/create_coverageplot.html">create_coverageplot</a></span><span class="op">(</span></span>
<span>  <span class="va">xhist</span>, </span>
<span>  add.points <span class="op">=</span> <span class="cn">T</span>, </span>
<span>  points.x <span class="op">=</span> <span class="va">optima_flat</span>, </span>
<span>  points.y <span class="op">=</span> <span class="va">data</span><span class="op">[</span><span class="va">optima_flat</span><span class="op">]</span>, </span>
<span>  points.col <span class="op">=</span> <span class="st">"red"</span>,</span>
<span>  main <span class="op">=</span> <span class="st">"flat endpoints"</span>,</span>
<span>  main.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  xaxis.tck <span class="op">=</span> <span class="fl">0</span>, </span>
<span>  yaxis.tck <span class="op">=</span> <span class="fl">0</span>, </span>
<span>  xlab.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  ylab.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  xaxis.cex <span class="op">=</span> <span class="fl">0.8</span>, </span>
<span>  yaxis.cex <span class="op">=</span> <span class="fl">0.8</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># midpoints</span></span>
<span><span class="va">optima_midpoints</span> <span class="op">=</span> <span class="fu"><a href="../reference/find_local_optima.html">find_local_optima</a></span><span class="op">(</span><span class="va">data</span>, flat_endpoints <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">optima_midpoints</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">optima_midpoints</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/create_coverageplot.html">create_coverageplot</a></span><span class="op">(</span></span>
<span>  <span class="va">xhist</span>, </span>
<span>  add.points <span class="op">=</span> <span class="cn">T</span>, </span>
<span>  points.x <span class="op">=</span> <span class="va">optima_midpoints</span>, </span>
<span>  points.y <span class="op">=</span> <span class="va">data</span><span class="op">[</span><span class="va">optima_midpoints</span><span class="op">]</span>, </span>
<span>  points.col <span class="op">=</span> <span class="st">"red"</span>,</span>
<span>  main <span class="op">=</span> <span class="st">"midpoints"</span>,</span>
<span>  main.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  xaxis.tck <span class="op">=</span> <span class="fl">0</span>, </span>
<span>  yaxis.tck <span class="op">=</span> <span class="fl">0</span>, </span>
<span>  xlab.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  ylab.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  xaxis.cex <span class="op">=</span> <span class="fl">0.8</span>, </span>
<span>  yaxis.cex <span class="op">=</span> <span class="fl">0.8</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="HistogramZoo_files/figure-html/find_local_optima_stepfunction-1.png" width="40%"><img src="HistogramZoo_files/figure-html/find_local_optima_stepfunction-2.png" width="40%"></p>
</div>
<div class="section level3">
<h3 id="the-fine-to-coarse-segmentation-algorithm">The Fine-to-Coarse Segmentation Algorithm<a class="anchor" aria-label="anchor" href="#the-fine-to-coarse-segmentation-algorithm"></a>
</h3>
<p>The Fine-to-Coarse segmentation algorithm was described in <span class="citation">Delon et al. (2005)</span>, <span class="citation">Lisani and Petro (2021)</span> and <span class="citation">Delon et al. (2006)</span>. Briefly, for a given
histogram and a set of local optima, the FTC algorithm iterates through
consecutive segments identifying <em>peaks</em>, segments of increasing
counts followed by segments of decreasing counts. Consecutive segments
are merged based on the prioritization derived from a monotone cost
function, until an optimal set of peaks are acquired. Granularity of the
optimal peak set can be tuned via the hyperparamter <span class="math inline">\(\epsilon\)</span>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Generating a Histogram with noisy counts</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">314</span><span class="op">)</span></span>
<span><span class="va">dt</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">1000</span>, mean <span class="op">=</span> <span class="fl">20</span>, sd <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">xhist</span> <span class="op">=</span> <span class="fu"><a href="../reference/observations_to_histogram.html">observations_to_histogram</a></span><span class="op">(</span><span class="va">dt</span>, histogram_bin_width <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extracting the histogram data</span></span>
<span><span class="va">data</span> <span class="op">=</span> <span class="va">xhist</span><span class="op">$</span><span class="va">histogram_data</span></span>
<span></span>
<span><span class="co"># local optima</span></span>
<span><span class="va">zero_threshold_optima</span> <span class="op">=</span> <span class="fu"><a href="../reference/find_local_optima.html">find_local_optima</a></span><span class="op">(</span><span class="va">data</span>, threshold <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">zero_threshold_optima</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">zero_threshold_optima</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/create_coverageplot.html">create_coverageplot</a></span><span class="op">(</span></span>
<span>  <span class="va">xhist</span>,</span>
<span>  add.points <span class="op">=</span> <span class="cn">T</span>, </span>
<span>  points.x <span class="op">=</span> <span class="va">zero_threshold_optima</span>,</span>
<span>  points.y <span class="op">=</span> <span class="va">data</span><span class="op">[</span><span class="va">zero_threshold_optima</span><span class="op">]</span>,</span>
<span>  points.col <span class="op">=</span> <span class="st">"red"</span>, </span>
<span>  type <span class="op">=</span> <span class="st">"h"</span>,</span>
<span>  main <span class="op">=</span> <span class="st">"Local Optima"</span>,</span>
<span>  main.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  xaxis.tck <span class="op">=</span> <span class="fl">0</span>, </span>
<span>  yaxis.tck <span class="op">=</span> <span class="fl">0</span>, </span>
<span>  xlab.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  ylab.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  xaxis.cex <span class="op">=</span> <span class="fl">0.8</span>, </span>
<span>  yaxis.cex <span class="op">=</span> <span class="fl">0.8</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Using different epsilons on FTC</span></span>
<span><span class="va">ftc_eps_0.05</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ftc.html">ftc</a></span><span class="op">(</span><span class="va">data</span>, s <span class="op">=</span> <span class="va">zero_threshold_optima</span>, eps <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="va">ftc_eps_0.09</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ftc.html">ftc</a></span><span class="op">(</span><span class="va">data</span>, s <span class="op">=</span> <span class="va">zero_threshold_optima</span>, eps <span class="op">=</span> <span class="fl">0.09</span><span class="op">)</span></span>
<span><span class="va">ftc_eps_0.15</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ftc.html">ftc</a></span><span class="op">(</span><span class="va">data</span>, s <span class="op">=</span> <span class="va">zero_threshold_optima</span>, eps <span class="op">=</span> <span class="fl">0.15</span><span class="op">)</span></span>
<span><span class="va">ftc_eps_0.25</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ftc.html">ftc</a></span><span class="op">(</span><span class="va">data</span>, s <span class="op">=</span> <span class="va">zero_threshold_optima</span>, eps <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span></span>
<span><span class="va">ftc_eps_0.5</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ftc.html">ftc</a></span><span class="op">(</span><span class="va">data</span>, s <span class="op">=</span> <span class="va">zero_threshold_optima</span>, eps <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/create_coverageplot.html">create_coverageplot</a></span><span class="op">(</span></span>
<span>  <span class="va">xhist</span>,</span>
<span>  add.points <span class="op">=</span> <span class="cn">T</span>, </span>
<span>  points.x <span class="op">=</span> <span class="va">ftc_eps_0.05</span>,</span>
<span>  points.y <span class="op">=</span> <span class="va">data</span><span class="op">[</span><span class="va">ftc_eps_0.05</span><span class="op">]</span>,</span>
<span>  points.col <span class="op">=</span> <span class="st">"red"</span>, </span>
<span>  type <span class="op">=</span> <span class="st">"h"</span>,</span>
<span>  main <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html" class="external-link">expression</a></span><span class="op">(</span><span class="fu">bold</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"FTC: "</span>, <span class="va">epsilon</span>, <span class="st">"=0.05"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  main.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  xaxis.tck <span class="op">=</span> <span class="fl">0</span>, </span>
<span>  yaxis.tck <span class="op">=</span> <span class="fl">0</span>, </span>
<span>  xlab.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  ylab.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  xaxis.cex <span class="op">=</span> <span class="fl">0.8</span>, </span>
<span>  yaxis.cex <span class="op">=</span> <span class="fl">0.8</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/create_coverageplot.html">create_coverageplot</a></span><span class="op">(</span></span>
<span>  <span class="va">xhist</span>,</span>
<span>  add.points <span class="op">=</span> <span class="cn">T</span>, </span>
<span>  points.x <span class="op">=</span> <span class="va">ftc_eps_0.09</span>,</span>
<span>  points.y <span class="op">=</span> <span class="va">data</span><span class="op">[</span><span class="va">ftc_eps_0.09</span><span class="op">]</span>,</span>
<span>  points.col <span class="op">=</span> <span class="st">"red"</span>, </span>
<span>  type <span class="op">=</span> <span class="st">"h"</span>,</span>
<span>  main <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html" class="external-link">expression</a></span><span class="op">(</span><span class="fu">bold</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"FTC: "</span>, <span class="va">epsilon</span>, <span class="st">"=0.09"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  main.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  xaxis.tck <span class="op">=</span> <span class="fl">0</span>, </span>
<span>  yaxis.tck <span class="op">=</span> <span class="fl">0</span>, </span>
<span>  xlab.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  ylab.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  xaxis.cex <span class="op">=</span> <span class="fl">0.8</span>, </span>
<span>  yaxis.cex <span class="op">=</span> <span class="fl">0.8</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/create_coverageplot.html">create_coverageplot</a></span><span class="op">(</span></span>
<span>  <span class="va">xhist</span>,</span>
<span>  add.points <span class="op">=</span> <span class="cn">T</span>, </span>
<span>  points.x <span class="op">=</span> <span class="va">ftc_eps_0.15</span>,</span>
<span>  points.y <span class="op">=</span> <span class="va">data</span><span class="op">[</span><span class="va">ftc_eps_0.15</span><span class="op">]</span>,</span>
<span>  points.col <span class="op">=</span> <span class="st">"red"</span>, </span>
<span>  type <span class="op">=</span> <span class="st">"h"</span>,</span>
<span>  main <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html" class="external-link">expression</a></span><span class="op">(</span><span class="fu">bold</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"FTC: "</span>, <span class="va">epsilon</span>, <span class="st">"=0.15"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  main.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  xaxis.tck <span class="op">=</span> <span class="fl">0</span>, </span>
<span>  yaxis.tck <span class="op">=</span> <span class="fl">0</span>, </span>
<span>  xlab.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  ylab.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  xaxis.cex <span class="op">=</span> <span class="fl">0.8</span>, </span>
<span>  yaxis.cex <span class="op">=</span> <span class="fl">0.8</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/create_coverageplot.html">create_coverageplot</a></span><span class="op">(</span></span>
<span>  <span class="va">xhist</span>,</span>
<span>  add.points <span class="op">=</span> <span class="cn">T</span>, </span>
<span>  points.x <span class="op">=</span> <span class="va">ftc_eps_0.25</span>,</span>
<span>  points.y <span class="op">=</span> <span class="va">data</span><span class="op">[</span><span class="va">ftc_eps_0.25</span><span class="op">]</span>,</span>
<span>  points.col <span class="op">=</span> <span class="st">"red"</span>, </span>
<span>  type <span class="op">=</span> <span class="st">"h"</span>,</span>
<span>  main <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html" class="external-link">expression</a></span><span class="op">(</span><span class="fu">bold</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"FTC: "</span>, <span class="va">epsilon</span>, <span class="st">"=0.25"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  main.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  xaxis.tck <span class="op">=</span> <span class="fl">0</span>, </span>
<span>  yaxis.tck <span class="op">=</span> <span class="fl">0</span>, </span>
<span>  xlab.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  ylab.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  xaxis.cex <span class="op">=</span> <span class="fl">0.8</span>, </span>
<span>  yaxis.cex <span class="op">=</span> <span class="fl">0.8</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/create_coverageplot.html">create_coverageplot</a></span><span class="op">(</span></span>
<span>  <span class="va">xhist</span>,</span>
<span>  add.points <span class="op">=</span> <span class="cn">T</span>, </span>
<span>  points.x <span class="op">=</span> <span class="va">ftc_eps_0.5</span>,</span>
<span>  points.y <span class="op">=</span> <span class="va">data</span><span class="op">[</span><span class="va">ftc_eps_0.5</span><span class="op">]</span>,</span>
<span>  points.col <span class="op">=</span> <span class="st">"red"</span>, </span>
<span>  type <span class="op">=</span> <span class="st">"h"</span>,</span>
<span>  main <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html" class="external-link">expression</a></span><span class="op">(</span><span class="fu">bold</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"FTC: "</span>, <span class="va">epsilon</span>, <span class="st">"=0.5"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  main.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  xaxis.tck <span class="op">=</span> <span class="fl">0</span>, </span>
<span>  yaxis.tck <span class="op">=</span> <span class="fl">0</span>, </span>
<span>  xlab.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  ylab.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  xaxis.cex <span class="op">=</span> <span class="fl">0.8</span>, </span>
<span>  yaxis.cex <span class="op">=</span> <span class="fl">0.8</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="HistogramZoo_files/figure-html/ftc_epsilon-1.png" width="40%"><img src="HistogramZoo_files/figure-html/ftc_epsilon-2.png" width="40%"><img src="HistogramZoo_files/figure-html/ftc_epsilon-3.png" width="40%"><img src="HistogramZoo_files/figure-html/ftc_epsilon-4.png" width="40%"><img src="HistogramZoo_files/figure-html/ftc_epsilon-5.png" width="40%"><img src="HistogramZoo_files/figure-html/ftc_epsilon-6.png" width="40%"></p>
</div>
<div class="section level3">
<h3 id="filtering-low-entropy-regions">Filtering Low Entropy Regions<a class="anchor" aria-label="anchor" href="#filtering-low-entropy-regions"></a>
</h3>
<p>In the presence of noisy data, it can be helpful to filter regions
with low entropy to extract high density regions. Low entropy regions or
<code>max_gaps</code> can be identified using the algorithm described in
<span class="citation">Delon et al. (2005)</span>. Briefly, a region is
defined as a gap if it has high entropy and low density relative to its
adjacent peak. Identification of max gaps is conducted on each segment
post-segmentation.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Generating a Histogram with noisy counts</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">314</span><span class="op">)</span></span>
<span><span class="va">dt</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">1000</span>, mean <span class="op">=</span> <span class="fl">20</span>, sd <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">1000</span>, min <span class="op">=</span> <span class="fl">0</span>, max <span class="op">=</span> <span class="fl">40</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">xhist</span> <span class="op">=</span> <span class="fu"><a href="../reference/observations_to_histogram.html">observations_to_histogram</a></span><span class="op">(</span><span class="va">dt</span>, histogram_bin_width <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Segment and Fit with and without filtering for max gaps</span></span>
<span><span class="va">filtered_fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/segment_and_fit.html">segment_and_fit</a></span><span class="op">(</span></span>
<span>  histogram_obj <span class="op">=</span> <span class="va">xhist</span>,</span>
<span>  eps <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">314</span>,</span>
<span>  remove_low_entropy <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  metric <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"jaccard"</span>, <span class="st">"intersection"</span>, <span class="st">"ks"</span>, <span class="st">"mse"</span>, <span class="st">"chisq"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">unfiltered_fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/segment_and_fit.html">segment_and_fit</a></span><span class="op">(</span></span>
<span>  histogram_obj <span class="op">=</span> <span class="va">xhist</span>,</span>
<span>  eps <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">314</span>,</span>
<span>  remove_low_entropy <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  metric <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"jaccard"</span>, <span class="st">"intersection"</span>, <span class="st">"ks"</span>, <span class="st">"mse"</span>, <span class="st">"chisq"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">unfiltered_plt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_coverageplot.html">create_coverageplot</a></span><span class="op">(</span></span>
<span>  <span class="va">unfiltered_fit</span>,</span>
<span>  main <span class="op">=</span> <span class="st">"Unfiltered"</span>,</span>
<span>  main.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  xaxis.tck <span class="op">=</span> <span class="fl">0</span>, </span>
<span>  yaxis.tck <span class="op">=</span> <span class="fl">0.5</span>, </span>
<span>  xlab.label <span class="op">=</span> <span class="st">''</span>, </span>
<span>  ylab.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  xaxis.lab <span class="op">=</span> <span class="cn">NULL</span>, </span>
<span>  yaxis.cex <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  legend <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plotting</span></span>
<span><span class="va">filtered_plt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_coverageplot.html">create_coverageplot</a></span><span class="op">(</span></span>
<span>  <span class="va">filtered_fit</span>,</span>
<span>  main <span class="op">=</span> <span class="st">"Remove max gaps"</span>,</span>
<span>  main.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  xaxis.tck <span class="op">=</span> <span class="fl">0</span>, </span>
<span>  yaxis.tck <span class="op">=</span> <span class="fl">0.5</span>, </span>
<span>  xlab.label <span class="op">=</span> <span class="st">''</span>, </span>
<span>  ylab.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  xaxis.label <span class="op">=</span> <span class="cn">NULL</span>, </span>
<span>  yaxis.cex <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  legend <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">unfiltered_plt</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">filtered_plt</span><span class="op">)</span></span></code></pre></div>
<p><img src="HistogramZoo_files/figure-html/low_entropy-1.png" width="40%"><img src="HistogramZoo_files/figure-html/low_entropy-2.png" width="40%"></p>
</div>
<div class="section level3">
<h3 id="fitting-statistical-distributions">Fitting Statistical Distributions<a class="anchor" aria-label="anchor" href="#fitting-statistical-distributions"></a>
</h3>
<div class="section level4">
<h4 id="distributions">Distributions<a class="anchor" aria-label="anchor" href="#distributions"></a>
</h4>
<p>Three distributions are provided: the normal (Gaussian) distribution,
the uniform distribution and the gamma distribution. The gamma
distribution can be flipped to account for skew in the negative
direction. Normal and gamma distributions can be fit in their full form
or in a truncated form.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">314</span><span class="op">)</span></span>
<span><span class="va">gaussian_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">10000</span>, mean <span class="op">=</span> <span class="fl">50</span>, sd <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">gaussian_histogram</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/observations_to_histogram.html">observations_to_histogram</a></span><span class="op">(</span><span class="va">gaussian_data</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Keeping all data and fitting a full model</span></span>
<span><span class="va">gaussian_full</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/segment_and_fit.html">segment_and_fit</a></span><span class="op">(</span></span>
<span>  histogram_obj <span class="op">=</span> <span class="va">gaussian_histogram</span>,</span>
<span>  eps <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">314</span>,</span>
<span>  remove_low_entropy <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  truncated_models <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  metric <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"jaccard"</span>, <span class="st">"intersection"</span>, <span class="st">"ks"</span>, <span class="st">"mse"</span>, <span class="st">"chisq"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Removing noisy data and fitting a truncated model</span></span>
<span><span class="va">gaussian_truncated</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/segment_and_fit.html">segment_and_fit</a></span><span class="op">(</span></span>
<span>  histogram_obj <span class="op">=</span> <span class="va">gaussian_histogram</span>,</span>
<span>  eps <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">314</span>,</span>
<span>  remove_low_entropy <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  truncated_models <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  metric <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"jaccard"</span>, <span class="st">"intersection"</span>, <span class="st">"ks"</span>, <span class="st">"mse"</span>, <span class="st">"chisq"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">gaussian_full_plt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_coverageplot.html">create_coverageplot</a></span><span class="op">(</span></span>
<span>  <span class="va">gaussian_full</span>,</span>
<span>  main <span class="op">=</span> <span class="st">"Normal (Full)"</span>,</span>
<span>  main.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  xaxis.tck <span class="op">=</span> <span class="fl">0</span>, </span>
<span>  yaxis.tck <span class="op">=</span> <span class="fl">0.5</span>, </span>
<span>  xlab.label <span class="op">=</span> <span class="st">''</span>, </span>
<span>  ylab.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  xaxis.label <span class="op">=</span> <span class="cn">NULL</span>, </span>
<span>  yaxis.cex <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  legend <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">gaussian_truncated_plt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_coverageplot.html">create_coverageplot</a></span><span class="op">(</span></span>
<span>  <span class="va">gaussian_truncated</span>,</span>
<span>  main <span class="op">=</span> <span class="st">"Normal (Truncated)"</span>,</span>
<span>  main.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  xaxis.tck <span class="op">=</span> <span class="fl">0</span>, </span>
<span>  yaxis.tck <span class="op">=</span> <span class="fl">0.5</span>, </span>
<span>  xlab.label <span class="op">=</span> <span class="st">''</span>, </span>
<span>  ylab.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  xaxis.label <span class="op">=</span> <span class="cn">NULL</span>, </span>
<span>  yaxis.cex <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  legend <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">gaussian_full_plt</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">gaussian_truncated_plt</span><span class="op">)</span></span></code></pre></div>
<p><img src="HistogramZoo_files/figure-html/distributions_normal-1.png" width="40%"><img src="HistogramZoo_files/figure-html/distributions_normal-2.png" width="40%"></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">314</span><span class="op">)</span></span>
<span><span class="va">gamma_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/GammaDist.html" class="external-link">rgamma</a></span><span class="op">(</span><span class="fl">10000</span>, shape <span class="op">=</span> <span class="fl">2</span>, rate <span class="op">=</span> <span class="fl">0.4</span><span class="op">)</span></span>
<span><span class="va">gamma_histogram</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/observations_to_histogram.html">observations_to_histogram</a></span><span class="op">(</span><span class="va">gamma_data</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Keeping all data and fitting a full model</span></span>
<span><span class="va">gamma_full</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/segment_and_fit.html">segment_and_fit</a></span><span class="op">(</span></span>
<span>  histogram_obj <span class="op">=</span> <span class="va">gamma_histogram</span>,</span>
<span>  eps <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">314</span>,</span>
<span>  remove_low_entropy <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  truncated_models <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  metric <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"jaccard"</span>, <span class="st">"intersection"</span>, <span class="st">"ks"</span>, <span class="st">"mse"</span>, <span class="st">"chisq"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Removing noisy data and fitting a truncated model</span></span>
<span><span class="va">gamma_truncated</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/segment_and_fit.html">segment_and_fit</a></span><span class="op">(</span></span>
<span>  histogram_obj <span class="op">=</span> <span class="va">gamma_histogram</span>,</span>
<span>  eps <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">314</span>,</span>
<span>  remove_low_entropy <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  truncated_models <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  metric <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"jaccard"</span>, <span class="st">"intersection"</span>, <span class="st">"ks"</span>, <span class="st">"mse"</span>, <span class="st">"chisq"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">gamma_full_plt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_coverageplot.html">create_coverageplot</a></span><span class="op">(</span></span>
<span>  <span class="va">gamma_full</span>,</span>
<span>  main <span class="op">=</span> <span class="st">"Gamma (Full)"</span>,</span>
<span>  main.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  xaxis.tck <span class="op">=</span> <span class="fl">0</span>, </span>
<span>  yaxis.tck <span class="op">=</span> <span class="fl">0.5</span>, </span>
<span>  xlab.label <span class="op">=</span> <span class="st">''</span>, </span>
<span>  ylab.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  xaxis.label <span class="op">=</span> <span class="cn">NULL</span>, </span>
<span>  yaxis.cex <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  legend <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">gamma_truncated_plt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_coverageplot.html">create_coverageplot</a></span><span class="op">(</span></span>
<span>  <span class="va">gamma_truncated</span>,</span>
<span>  main <span class="op">=</span> <span class="st">"Gamma (Truncated)"</span>,</span>
<span>  main.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  xaxis.tck <span class="op">=</span> <span class="fl">0</span>, </span>
<span>  yaxis.tck <span class="op">=</span> <span class="fl">0.5</span>, </span>
<span>  xlab.label <span class="op">=</span> <span class="st">''</span>, </span>
<span>  ylab.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  xaxis.label <span class="op">=</span> <span class="cn">NULL</span>, </span>
<span>  yaxis.cex <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  legend <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">gamma_full_plt</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">gamma_truncated_plt</span><span class="op">)</span></span></code></pre></div>
<p><img src="HistogramZoo_files/figure-html/distributions_gamma-1.png" width="40%"><img src="HistogramZoo_files/figure-html/distributions_gamma-2.png" width="40%"></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">314</span><span class="op">)</span></span>
<span><span class="va">gamma_flip_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rgamma_flip.html">rgamma_flip</a></span><span class="op">(</span><span class="fl">9000</span>, shape <span class="op">=</span> <span class="fl">4</span>, rate <span class="op">=</span> <span class="fl">0.4</span><span class="op">)</span> <span class="op">+</span> <span class="fl">80</span></span>
<span><span class="va">gamma_flip_histogram</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/observations_to_histogram.html">observations_to_histogram</a></span><span class="op">(</span><span class="va">gamma_flip_data</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Keeping all data and fitting a full model</span></span>
<span><span class="va">gamma_flip_full</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/segment_and_fit.html">segment_and_fit</a></span><span class="op">(</span></span>
<span>  histogram_obj <span class="op">=</span> <span class="va">gamma_flip_histogram</span>,</span>
<span>  eps <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">314</span>,</span>
<span>  remove_low_entropy <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  truncated_models <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  metric <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"jaccard"</span>, <span class="st">"intersection"</span>, <span class="st">"ks"</span>, <span class="st">"mse"</span>, <span class="st">"chisq"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Removing noisy data and fitting a truncated model</span></span>
<span><span class="va">gamma_flip_truncated</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/segment_and_fit.html">segment_and_fit</a></span><span class="op">(</span></span>
<span>  histogram_obj <span class="op">=</span> <span class="va">gamma_flip_histogram</span>,</span>
<span>  eps <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">314</span>,</span>
<span>  remove_low_entropy <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  truncated_models <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  metric <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"jaccard"</span>, <span class="st">"intersection"</span>, <span class="st">"ks"</span>, <span class="st">"mse"</span>, <span class="st">"chisq"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">gamma_flip_full_plt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_coverageplot.html">create_coverageplot</a></span><span class="op">(</span></span>
<span>  <span class="va">gamma_flip_full</span>,</span>
<span>  main <span class="op">=</span> <span class="st">"Gamma Flipped (Full)"</span>,</span>
<span>  main.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  xaxis.tck <span class="op">=</span> <span class="fl">0</span>, </span>
<span>  yaxis.tck <span class="op">=</span> <span class="fl">0.5</span>, </span>
<span>  xlab.label <span class="op">=</span> <span class="st">''</span>, </span>
<span>  ylab.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  xaxis.label <span class="op">=</span> <span class="cn">NULL</span>, </span>
<span>  yaxis.cex <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  legend <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">gamma_flip_truncated_plt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_coverageplot.html">create_coverageplot</a></span><span class="op">(</span></span>
<span>  <span class="va">gamma_flip_truncated</span>,</span>
<span>  main <span class="op">=</span> <span class="st">"Gamma Flipped (Truncated)"</span>,</span>
<span>  main.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  xaxis.tck <span class="op">=</span> <span class="fl">0</span>, </span>
<span>  yaxis.tck <span class="op">=</span> <span class="fl">0.5</span>, </span>
<span>  xlab.label <span class="op">=</span> <span class="st">''</span>, </span>
<span>  ylab.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  xaxis.label <span class="op">=</span> <span class="cn">NULL</span>, </span>
<span>  yaxis.cex <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  legend <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">gamma_flip_full_plt</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">gamma_flip_truncated_plt</span><span class="op">)</span></span></code></pre></div>
<p><img src="HistogramZoo_files/figure-html/distributions_gamma_flip-1.png" width="40%"><img src="HistogramZoo_files/figure-html/distributions_gamma_flip-2.png" width="40%"></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">314</span><span class="op">)</span></span>
<span><span class="va">uniform_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">10000</span>, min <span class="op">=</span> <span class="fl">0</span>, max <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">uniform_histogram</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/observations_to_histogram.html">observations_to_histogram</a></span><span class="op">(</span><span class="va">uniform_data</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Fitting a uniform segment</span></span>
<span><span class="va">uniform_fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/segment_and_fit.html">segment_and_fit</a></span><span class="op">(</span></span>
<span>  histogram_obj <span class="op">=</span> <span class="va">uniform_histogram</span>,</span>
<span>  eps <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">314</span>,</span>
<span>  max_uniform <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  metric <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"jaccard"</span>, <span class="st">"intersection"</span>, <span class="st">"ks"</span>, <span class="st">"mse"</span>, <span class="st">"chisq"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">uniform_plt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_coverageplot.html">create_coverageplot</a></span><span class="op">(</span></span>
<span>  <span class="va">uniform_fit</span>,</span>
<span>  main <span class="op">=</span> <span class="st">"Uniform"</span>,</span>
<span>  main.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  xaxis.tck <span class="op">=</span> <span class="fl">0</span>, </span>
<span>  yaxis.tck <span class="op">=</span> <span class="fl">0.5</span>, </span>
<span>  xlab.label <span class="op">=</span> <span class="st">''</span>, </span>
<span>  ylab.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  xaxis.label <span class="op">=</span> <span class="cn">NULL</span>, </span>
<span>  yaxis.cex <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  legend <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">uniform_plt</span><span class="op">)</span></span></code></pre></div>
<p><img src="HistogramZoo_files/figure-html/distributions_uniform-1.png" width="40%"></p>
</div>
<div class="section level4">
<h4 id="metrics">Metrics<a class="anchor" aria-label="anchor" href="#metrics"></a>
</h4>
<p>Five metrics are provided to optimize for the best fitting
statistical distribution.</p>
<ul>
<li>
<strong>Jaccard index</strong> is the ratio between the intersection
of the fitted and observed distributions and the union of the two
distributions.</li>
<li>
<strong>Intersection</strong> is the intersection between the fitted
and observed distributions.</li>
<li>
<strong>Kolmogorov-Smirnov statistic</strong> is the maximum
divergence between the fitted and observed distributions.</li>
<li>
<strong>Mean squared error</strong> is the mean squared difference
between the fitted and observed distributions.</li>
<li>
<strong><span class="math inline">\(\chi\)</span>-squared
statistic</strong> is the <span class="math inline">\(\chi\)</span>-squared distance between the two
distributions i.e. <span class="math inline">\(\sum{\frac{(x-y)^2}{(x+y)}}\)</span>
</li>
</ul>
</div>
<div class="section level4">
<h4 id="finding-a-consensus-model">Finding a Consensus Model<a class="anchor" aria-label="anchor" href="#finding-a-consensus-model"></a>
</h4>
<p>The function <code>find_consensus_model</code> is used to determine
the best fitting model for each segment and provides two methods of
doing so: <code>weighted_majority_vote</code> (weighted majority voting)
and <code>rra</code> (robust rank aggregation, <span class="citation">Kolde et al. (2012)</span>). Weighted majority voting
requires the users to provide a ranking of metrics in descending order
of importance and their associated weights and used the weighted sum of
ranks to determine the best fitting distribution. In the event of a tie,
the highest ranked metric is used to break the tie.</p>
</div>
<div class="section level4">
<h4 id="maximizing-uniform-segments">Maximizing Uniform Segments<a class="anchor" aria-label="anchor" href="#maximizing-uniform-segments"></a>
</h4>
<p>The function <code>identify_uniform_segment</code> optimizes the
identification of uniform segments by iteratively trimming the ends of
the region and refitting a uniform distribution to the shortened region.
This is controlled by the <code>threshold</code> parameter which is the
proportion of the original region which must be preserved and the
<code>stepsize</code> which indicates the stepsize taken for each
search. One of the metrics must be selected to determine the best fit
while the parameter <code>max_sd_size</code> indicates the maximum
tolerable standard deviation of the goodness-of-fit from the best
fitting uniform distribution where the function can select the longest
uniform distribution.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">314</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Simulating uniform data that can benefit from trimming</span></span>
<span><span class="va">uniform_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">10000</span>, min <span class="op">=</span> <span class="fl">0</span>, max <span class="op">=</span> <span class="fl">50</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">10000</span>, min <span class="op">=</span> <span class="fl">2</span>, max <span class="op">=</span> <span class="fl">48</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">10000</span>, min <span class="op">=</span> <span class="fl">4</span>, max <span class="op">=</span> <span class="fl">46</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">10000</span>, min <span class="op">=</span> <span class="fl">6</span>, max <span class="op">=</span> <span class="fl">44</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">10000</span>, min <span class="op">=</span> <span class="fl">8</span>, max <span class="op">=</span> <span class="fl">42</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">uniform_histogram</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/observations_to_histogram.html">observations_to_histogram</a></span><span class="op">(</span><span class="va">uniform_data</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Fitting a uniform segment</span></span>
<span><span class="va">uniform_untrimmed</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/segment_and_fit.html">segment_and_fit</a></span><span class="op">(</span></span>
<span>  histogram_obj <span class="op">=</span> <span class="va">uniform_histogram</span>,</span>
<span>  eps <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">314</span>,</span>
<span>  max_uniform <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  remove_low_entropy <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  metric <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"jaccard"</span>, <span class="st">"intersection"</span>, <span class="st">"ks"</span>, <span class="st">"mse"</span>, <span class="st">"chisq"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">uniform_trimmed</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/segment_and_fit.html">segment_and_fit</a></span><span class="op">(</span></span>
<span>  histogram_obj <span class="op">=</span> <span class="va">uniform_histogram</span>,</span>
<span>  eps <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">314</span>,</span>
<span>  max_uniform <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  uniform_threshold <span class="op">=</span> <span class="fl">0.75</span>,</span>
<span>  uniform_stepsize <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  uniform_max_sd <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  remove_low_entropy <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  metric <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"jaccard"</span>, <span class="st">"intersection"</span>, <span class="st">"ks"</span>, <span class="st">"mse"</span>, <span class="st">"chisq"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">uniform_untrimmed_plt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_coverageplot.html">create_coverageplot</a></span><span class="op">(</span></span>
<span>  <span class="va">uniform_untrimmed</span>,</span>
<span>  main <span class="op">=</span> <span class="st">"Uniform (Untrimmed)"</span>,</span>
<span>  main.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  xaxis.tck <span class="op">=</span> <span class="fl">0</span>, </span>
<span>  yaxis.tck <span class="op">=</span> <span class="fl">0.5</span>, </span>
<span>  xlab.label <span class="op">=</span> <span class="st">''</span>, </span>
<span>  ylab.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  xaxis.label <span class="op">=</span> <span class="cn">NULL</span>, </span>
<span>  yaxis.cex <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  legend <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">uniform_trimmed_plt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_coverageplot.html">create_coverageplot</a></span><span class="op">(</span></span>
<span>  <span class="va">uniform_trimmed</span>,</span>
<span>  main <span class="op">=</span> <span class="st">"Uniform (Max Uniform)"</span>,</span>
<span>  main.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  xaxis.tck <span class="op">=</span> <span class="fl">0</span>, </span>
<span>  yaxis.tck <span class="op">=</span> <span class="fl">0.5</span>, </span>
<span>  xlab.label <span class="op">=</span> <span class="st">''</span>, </span>
<span>  ylab.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  xaxis.label <span class="op">=</span> <span class="cn">NULL</span>, </span>
<span>  yaxis.cex <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  legend <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">uniform_untrimmed_plt</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">uniform_trimmed_plt</span><span class="op">)</span></span></code></pre></div>
<p><img src="HistogramZoo_files/figure-html/identify_uniform_segment-1.png" width="40%"><img src="HistogramZoo_files/figure-html/identify_uniform_segment-2.png" width="40%"></p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="exporting-results">Exporting Results<a class="anchor" aria-label="anchor" href="#exporting-results"></a>
</h2>
<div class="section level3">
<h3 id="summary-statistics">Summary statistics<a class="anchor" aria-label="anchor" href="#summary-statistics"></a>
</h3>
<p><strong>summarize_results</strong> can be applied to a
<code>HistogramFit</code> object to extract the indices and summary
statistics of the analysis.</p>
<p>A data frame will be returned with the following columns.</p>
<ul>
<li>
<strong>region_id</strong> character string denoting the region_id
of the Histogram</li>
<li>
<strong>segment_id</strong> an integer id identifying the ordinal
segment of the Histogram segmentation</li>
<li>
<strong>chr</strong> an optional column denoting the chromosome of a
GenomicHistogram object</li>
<li>
<strong>start</strong> the interval start of the segment</li>
<li>
<strong>end</strong> the interval end of the segment</li>
<li>
<strong>strand</strong> an optional column denoting the strand of a
GenomicHistogram object</li>
<li>
<strong>interval_count</strong> the number of intervals in the
segment - used for collapsing disjoint intervals</li>
<li>
<strong>interval_sizes</strong> the width of each interval</li>
<li>
<strong>interval_starts</strong> the start index of each
interval</li>
<li>
<strong>histogram_start</strong> The start index of the segment in
the Histogram representation</li>
<li>
<strong>histogram_end</strong> The end index of the segment in the
Histogram representation</li>
<li>
<strong>value</strong> the fitted value of the metric function</li>
<li>
<strong>metric</strong> the metric used to fit the distribution</li>
<li>
<strong>dist</strong> the distribution name</li>
<li>
<strong>dist_param[0-9]</strong> the values of distribution
parameters</li>
<li>
<strong>dist_param_name[0-9]</strong> the matching names of
distribution parameters</li>
</ul>
</div>
<div class="section level3">
<h3 id="plotting">Plotting<a class="anchor" aria-label="anchor" href="#plotting"></a>
</h3>
<p>Several plotting functions are available to aid with the
visualization of results. Plotting functions can be combined with
plotting capabilities from <code>BoutrosLab.plotting.general</code>
<span class="citation">(P’ng et al. 2019)</span> based on the
<code>lattice</code> R package <span class="citation">(Sarkar, Sarkar,
and KernSmooth 2015)</span>.</p>
<ul>
<li><p><strong>create_coverageplot</strong> can be used to visualize the
<code>Histogram</code>, <code>GenomicHistogram</code> or
<code>HistogramFit</code> object with annotated segmentation points and
distributions.</p></li>
<li><p><strong>create_residualplot</strong> can be used to visualize the
residuals between the fitted and observed distributions of a
<code>HistogramFit</code> object</p></li>
<li><p><strong>create_trackplot</strong> can be used to add
complementary tracks to a histogram visualization when combined with
<code>create_layerplot</code></p></li>
<li><p><strong>create_layerplot</strong> can be used to create a
multi-panel plot with combinations of coverage, residual and track
plots.</p></li>
</ul>
</div>
<div class="section level3">
<h3 id="example">Example<a class="anchor" aria-label="anchor" href="#example"></a>
</h3>
<p>A histogram is simulated from 3 statistical distributions.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">314</span><span class="op">)</span></span>
<span><span class="va">my_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">8000</span>, mean <span class="op">=</span> <span class="fl">50</span>, sd <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/GammaDist.html" class="external-link">rgamma</a></span><span class="op">(</span><span class="fl">10000</span>, shape <span class="op">=</span> <span class="fl">2</span>, rate <span class="op">=</span> <span class="fl">0.4</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="../reference/rgamma_flip.html">rgamma_flip</a></span><span class="op">(</span><span class="fl">9000</span>, shape <span class="op">=</span> <span class="fl">2</span>, rate <span class="op">=</span> <span class="fl">0.4</span><span class="op">)</span> <span class="op">+</span> <span class="fl">80</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">my_histogram</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/observations_to_histogram.html">observations_to_histogram</a></span><span class="op">(</span><span class="va">my_data</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># A basic coverageplot to visualize the data</span></span>
<span><span class="fu"><a href="../reference/create_coverageplot.html">create_coverageplot</a></span><span class="op">(</span></span>
<span>  <span class="va">my_histogram</span>,</span>
<span>  main <span class="op">=</span> <span class="st">"Histogram"</span>,</span>
<span>  main.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  xaxis.tck <span class="op">=</span> <span class="fl">0.5</span>, </span>
<span>  yaxis.tck <span class="op">=</span> <span class="fl">0.5</span>, </span>
<span>  xlab.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  ylab.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  xaxis.cex <span class="op">=</span> <span class="fl">0.8</span>, </span>
<span>  yaxis.cex <span class="op">=</span> <span class="fl">0.8</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="HistogramZoo_files/figure-html/plotting_example_p1-1.png" width="50%"></p>
<p>HistogramZoo is applied to segment the histogram and fit
distributions. Results are reported in a table.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Conducting analysis</span></span>
<span><span class="va">histogram_fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/segment_and_fit.html">segment_and_fit</a></span><span class="op">(</span></span>
<span>  histogram_obj <span class="op">=</span> <span class="va">my_histogram</span>,</span>
<span>  eps <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">314</span>,</span>
<span>  truncated_models <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  metric <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"jaccard"</span>, <span class="st">"intersection"</span>, <span class="st">"ks"</span>, <span class="st">"mse"</span>, <span class="st">"chisq"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># summarize_results to generate</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/summarize_results.html">summarize_results</a></span><span class="op">(</span></span>
<span>  <span class="va">histogram_fit</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The <code>summarize_results</code> table returns a data frame with
columns describing the <code>region_id</code> of the original histogram
and a separate for each identified segment (<code>segment_id</code>).
The <code>start</code> and <code>end</code> columns indicate the segment
indices relative to the <code>interval_start</code> and
<code>interval_end</code> indices of the original histogram. The
<code>interval_count</code>, <code>interval_sizes</code> and
<code>interval_starts</code> columns aid in describing histograms with
disjoint intervals following a <a href="https://genome.ucsc.edu/FAQ/FAQformat.html" class="external-link">BED file
format</a>.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">res</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"region_id"</span>, <span class="st">"segment_id"</span>, <span class="st">"start"</span>, <span class="st">"end"</span>, <span class="st">"interval_count"</span>, <span class="st">"interval_sizes"</span>, <span class="st">"interval_starts"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="12%">
<col width="14%">
<col width="7%">
<col width="5%">
<col width="19%">
<col width="19%">
<col width="20%">
</colgroup>
<thead><tr class="header">
<th align="left">region_id</th>
<th align="right">segment_id</th>
<th align="right">start</th>
<th align="right">end</th>
<th align="right">interval_count</th>
<th align="left">interval_sizes</th>
<th align="left">interval_starts</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">0-80</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">10</td>
<td align="right">1</td>
<td align="left">11,</td>
<td align="left">1,</td>
</tr>
<tr class="even">
<td align="left">0-80</td>
<td align="right">2</td>
<td align="right">46</td>
<td align="right">54</td>
<td align="right">1</td>
<td align="left">9,</td>
<td align="left">1,</td>
</tr>
<tr class="odd">
<td align="left">0-80</td>
<td align="right">3</td>
<td align="right">71</td>
<td align="right">80</td>
<td align="right">1</td>
<td align="left">10,</td>
<td align="left">1,</td>
</tr>
</tbody>
</table>
<p>The <code>histogram_start</code> and <code>histogram_end</code>
columns indicate the start and end indices relative to the histogram
itself (i.e. assuming that each bin is ordinal). The <code>value</code>
refers to the GOF of the selected <code>metric</code> and the
<code>dist</code> column indicates the distribution fitted.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">res</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"histogram_start"</span>, <span class="st">"histogram_end"</span>, <span class="st">"value"</span>, <span class="st">"metric"</span>, <span class="st">"dist"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="right">histogram_start</th>
<th align="right">histogram_end</th>
<th align="right">value</th>
<th align="left">metric</th>
<th align="left">dist</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="right">1</td>
<td align="right">10</td>
<td align="right">0.9560695</td>
<td align="left">consensus</td>
<td align="left">gamma</td>
</tr>
<tr class="even">
<td align="right">47</td>
<td align="right">54</td>
<td align="right">0.9619485</td>
<td align="left">consensus</td>
<td align="left">norm</td>
</tr>
<tr class="odd">
<td align="right">72</td>
<td align="right">80</td>
<td align="right">0.9473543</td>
<td align="left">consensus</td>
<td align="left">gamma_flip</td>
</tr>
</tbody>
</table>
<p>The <code>dist_param</code> columns hold the values of the
distribution parameters</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">res</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"dist_param1"</span>, <span class="st">"dist_param2"</span>, <span class="st">"dist_param3"</span>, <span class="st">"dist_param4"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="right">dist_param1</th>
<th align="right">dist_param2</th>
<th align="right">dist_param3</th>
<th align="right">dist_param4</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="right">2.678828</td>
<td align="right">0.4942402</td>
<td align="right">1</td>
<td align="right">10</td>
</tr>
<tr class="even">
<td align="right">4.527836</td>
<td align="right">1.9906070</td>
<td align="right">1</td>
<td align="right">8</td>
</tr>
<tr class="odd">
<td align="right">2.670898</td>
<td align="right">0.4949953</td>
<td align="right">1</td>
<td align="right">9</td>
</tr>
</tbody>
</table>
<p>while the <code>dist_param_name</code> columns hold the names of the
parameters corresponding to each distribution</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">res</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"dist_param1_name"</span>, <span class="st">"dist_param2_name"</span>, <span class="st">"dist_param3_name"</span>, <span class="st">"dist_param4_name"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="25%">
<col width="25%">
<col width="25%">
<col width="25%">
</colgroup>
<thead><tr class="header">
<th align="left">dist_param1_name</th>
<th align="left">dist_param2_name</th>
<th align="left">dist_param3_name</th>
<th align="left">dist_param4_name</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">shape</td>
<td align="left">rate</td>
<td align="left">a</td>
<td align="left">b</td>
</tr>
<tr class="even">
<td align="left">mean</td>
<td align="left">sd</td>
<td align="left">a</td>
<td align="left">b</td>
</tr>
<tr class="odd">
<td align="left">shape</td>
<td align="left">rate</td>
<td align="left">a</td>
<td align="left">b</td>
</tr>
</tbody>
</table>
<p>A plot of the result is generated.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Histogram Fit</span></span>
<span><span class="va">coverage_plt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_coverageplot.html">create_coverageplot</a></span><span class="op">(</span></span>
<span>  <span class="va">histogram_fit</span>,</span>
<span>  main <span class="op">=</span> <span class="st">"HistogramFit"</span>,</span>
<span>  main.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  xaxis.tck <span class="op">=</span> <span class="fl">0</span>, </span>
<span>  yaxis.tck <span class="op">=</span> <span class="fl">0.5</span>, </span>
<span>  xlab.label <span class="op">=</span> <span class="st">''</span>, </span>
<span>  ylab.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  xaxis.label <span class="op">=</span> <span class="cn">NULL</span>, </span>
<span>  yaxis.cex <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  legend <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">residual_plt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_residualplot.html">create_residualplot</a></span><span class="op">(</span></span>
<span>  <span class="va">histogram_fit</span>,</span>
<span>  main <span class="op">=</span> <span class="st">''</span>, </span>
<span>  xaxis.tck <span class="op">=</span> <span class="fl">0</span>, </span>
<span>  yaxis.tck <span class="op">=</span> <span class="fl">0.5</span>, </span>
<span>  xlab.label <span class="op">=</span> <span class="st">''</span>, </span>
<span>  ylab.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  ylab.label <span class="op">=</span> <span class="st">"Residuals"</span>,</span>
<span>  xaxis.label <span class="op">=</span> <span class="cn">NULL</span>, </span>
<span>  yaxis.cex <span class="op">=</span> <span class="fl">0.8</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">track_plt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_trackplot.html">create_trackplot</a></span><span class="op">(</span></span>
<span>  <span class="va">res</span>,</span>
<span>  row_id <span class="op">=</span> <span class="st">"region_id"</span>,</span>
<span>  metric_id <span class="op">=</span> <span class="st">"value"</span>,</span>
<span>  start_id <span class="op">=</span> <span class="st">"histogram_start"</span>,</span>
<span>  end_id <span class="op">=</span> <span class="st">"histogram_end"</span>,</span>
<span>  xat <span class="op">=</span> <span class="fu"><a href="../reference/generate_xlabels.html">generate_xlabels</a></span><span class="op">(</span><span class="va">histogram_fit</span>, return_xat <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>,</span>
<span>  xaxis.lab <span class="op">=</span> <span class="fu"><a href="../reference/generate_xlabels.html">generate_xlabels</a></span><span class="op">(</span><span class="va">histogram_fit</span><span class="op">)</span>,</span>
<span>  main <span class="op">=</span> <span class="st">''</span>, </span>
<span>  xaxis.tck <span class="op">=</span> <span class="fl">0</span>, </span>
<span>  yaxis.tck <span class="op">=</span> <span class="fl">0</span>, </span>
<span>  xlab.cex <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  ylab.cex <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  ylab.label <span class="op">=</span> <span class="st">"GOF"</span>,</span>
<span>  xaxis.cex <span class="op">=</span> <span class="fl">0.8</span>, </span>
<span>  yaxis.lab <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">covariate.legend</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    colours <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>      <span class="st">"orange"</span>,</span>
<span>      <span class="st">"chartreuse4"</span>,</span>
<span>      <span class="st">"chartreuse3"</span></span>
<span>    <span class="op">)</span>,</span>
<span>    labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>      <span class="st">"Normal"</span>,</span>
<span>      <span class="st">"Gamma"</span>,</span>
<span>      <span class="st">"Gamma Flipped"</span></span>
<span>    <span class="op">)</span>,</span>
<span>    title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html" class="external-link">expression</a></span><span class="op">(</span><span class="fu">bold</span><span class="op">(</span><span class="fu">underline</span><span class="op">(</span><span class="st">'Fitted Distributions'</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    lwd <span class="op">=</span> <span class="fl">0.5</span></span>
<span>  <span class="op">)</span>,</span>
<span>  legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    colours <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"red"</span><span class="op">)</span>,</span>
<span>    labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"FTC Segmentation"</span><span class="op">)</span>,</span>
<span>    title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html" class="external-link">expression</a></span><span class="op">(</span><span class="fu">bold</span><span class="op">(</span><span class="fu">underline</span><span class="op">(</span><span class="st">'Points'</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    lwd <span class="op">=</span> <span class="fl">0.5</span></span>
<span>  <span class="op">)</span>,</span>
<span>  legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    colours <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'blue'</span>, <span class="st">'red'</span><span class="op">)</span>,</span>
<span>    labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/formatc.html" class="external-link">formatC</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">res</span><span class="op">[</span>,<span class="st">"value"</span><span class="op">]</span><span class="op">)</span>, digits <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>,</span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/formatc.html" class="external-link">formatC</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">res</span><span class="op">[</span>,<span class="st">"value"</span><span class="op">]</span><span class="op">)</span>, digits <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span>    <span class="op">)</span>,</span>
<span>    title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html" class="external-link">expression</a></span><span class="op">(</span><span class="fu">bold</span><span class="op">(</span><span class="fu">underline</span><span class="op">(</span><span class="st">'Fit'</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    continuous <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    width <span class="op">=</span> <span class="fl">3</span>,</span>
<span>    tck <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    tck.number <span class="op">=</span> <span class="fl">3</span>,</span>
<span>    at <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">100</span><span class="op">)</span>,</span>
<span>    angle <span class="op">=</span> <span class="op">-</span><span class="fl">90</span>,</span>
<span>    just <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"center"</span>,<span class="st">"bottom"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">side.legend</span> <span class="op">&lt;-</span> <span class="fu">legend.grob</span><span class="op">(</span></span>
<span>  legends <span class="op">=</span> <span class="va">covariate.legend</span>,</span>
<span>  label.cex <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  title.cex <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  title.just <span class="op">=</span> <span class="st">'left'</span>,</span>
<span>  title.fontface <span class="op">=</span> <span class="st">'bold'</span>,</span>
<span>  size <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">summary_figure</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_layerplot.html">create_layerplot</a></span><span class="op">(</span></span>
<span>  plot.objects <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">coverage_plt</span>, <span class="va">residual_plt</span>, <span class="va">track_plt</span><span class="op">)</span>,</span>
<span>  plot.objects.heights <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4</span>,<span class="fl">1.5</span>,<span class="fl">1.5</span><span class="op">)</span>,</span>
<span>  y.spacing <span class="op">=</span> <span class="op">-</span><span class="fl">3.5</span>,</span>
<span>  ylab.axis.padding <span class="op">=</span> <span class="op">-</span><span class="fl">13</span>,</span>
<span>  legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    right <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>      x <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>      y <span class="op">=</span> <span class="fl">1</span>,</span>
<span>      fun <span class="op">=</span> <span class="va">side.legend</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  right.legend.padding <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  right.padding <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># For exporting the plot</span></span>
<span><span class="co"># filename = "SummaryFigure.pdf"</span></span>
<span><span class="co"># pdf(filename, width = 7, height = 7)</span></span>
<span><span class="co"># print(summary_figure)</span></span>
<span><span class="co"># dev.off()</span></span></code></pre></div>
<div class="figure">
<img src="SummaryFigure.jpg" alt="Summary Figure" width="80%"><p class="caption">
Summary Figure
</p>
</div>
</div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-delon2005color" class="csl-entry">
Delon, Julie, Agnes Desolneux, Jose Luis Lisani, and Ana Belen Petro.
2005. <span>“Color Image Segmentation Using Acceptable Histogram
Segmentation.”</span> In <em>Iberian Conference on Pattern Recognition
and Image Analysis</em>, 239–46. Springer.
</div>
<div id="ref-delon2006nonparametric" class="csl-entry">
Delon, Julie, Agns Desolneux, Jos-Luis Lisani, and Ana Beln Petro. 2006.
<span>“A Nonparametric Approach for Histogram Segmentation.”</span>
<em>IEEE Transactions on Image Processing</em> 16 (1): 253–61.
</div>
<div id="ref-kolde2012robust" class="csl-entry">
Kolde, Raivo, Sven Laur, Priit Adler, and Jaak Vilo. 2012. <span>“Robust
Rank Aggregation for Gene List Integration and Meta-Analysis.”</span>
<em>Bioinformatics</em> 28 (4): 573–80.
</div>
<div id="ref-lisani2021automatic" class="csl-entry">
Lisani, Jose Luis, and Ana Belén Petro. 2021. <span>“Automatic 1D
Histogram Segmentation and Application to the Computation of Color
Palettes.”</span>
</div>
<div id="ref-p2019bpg" class="csl-entry">
P’ng, Christine, Jeffrey Green, Lauren C Chong, Daryl Waggott, Stephenie
D Prokopec, Mehrdad Shamsi, Francis Nguyen, et al. 2019. <span>“BPG:
Seamless, Automated and Interactive Visualization of Scientific
Data.”</span> <em>BMC Bioinformatics</em> 20 (1): 1–5.
</div>
<div id="ref-sarkar2015package" class="csl-entry">
Sarkar, Deepayan, Maintainer Deepayan Sarkar, and Suggests KernSmooth.
2015. <span>“Package <span>‘Lattice’</span>.”</span> <em>Version
0.20</em> 33.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Helen Zhu, Stefan Eng.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
